File: ch01/E01B_frequent_words.py
# Frequent Words Problem:
# Find the most frequent k-mers in a string
#   Input: A string Text and an integer k
#   Output: All most frequent k-mers in Text

from utils import Text
from ch01.E01A_pattern_count import PatternCount


# Complexity:
# - Each iteration of first cycle: (n - k + 1) * k, because of PatternCount
# - First cycle: n - k + 1
# - Second cycle: n - k + 1
# O(n - k + 1) * O(k) * O(n - k + 1) + O(n - k + 1) ~ O(n^2 * k)
def FrequentWords(genome, k, limit=None):
    frequent_patterns = set()
    count = [0] * len(genome)

    # For every index, get pattern of length k and count how many times the pattern appears in the string
    for i in range(len(genome) - k):
        pattern = Text(genome, i, k)
        count[i], _ = PatternCount(genome, pattern)

    if limit is None:
        # Find the maximum pattern count
        max_count = max(count)

        # Cycle through the list and append to set of frequent patterns the ones whose count equals the maximum count
        for i in range(len(genome) - k):
            if count[i] == max_count:
                frequent_patterns.add(Text(genome, i, k))

    else:
        # Cycle through the list and append to set of frequent patterns the ones whose count is more than the limit
        for i in range(len(genome) - k):
            if count[i] >= limit:
                frequent_patterns.add(Text(genome, i, k))

    return frequent_patterns


def FastFrequentWords(genome, k):
    freq_patterns = set()
    freq_arr = ComputeFrequencies(genome, k)
    max_count = max(freq_arr)

    for i in range(pow(4, k)):
        if freq_arr[i] == max_count:
            pattern = NumberToPattern(i, k)
            freq_patterns.add(pattern)

    return freq_patterns


def FastFrequentWordsBySorting(genome, k):
    freq_patterns = set()
    index = [0] * (len(genome) - k)
    count = [0] * (len(genome) - k)

    for i in range(len(genome) - k):
        pattern = Text(genome, i, k)
        index[i] = PatternToNumber(pattern)
        count[i] = 1

    index = sorted(index)

    for i in range(1, len(genome) - k):
        if index[i] == index[i-1]:
            count[i] = count[i-1] + 1

    max_count = max(count)

    for i in range(len(genome) - k):
        if count[i] == max_count:
            pattern = NumberToPattern(index[i], k)
            freq_patterns.add(pattern)

    return freq_patterns


if __name__ == "__main__":
    _genome = input("Genome: ").upper()
    _k = int(input("k: "))

    for word in FrequentWords(_genome, _k):
        print(word)

File: cpp/ch01/E01B_frequent_words.cpp
//
// Created by Matej Hrnčiar on 20/07/2022.
//

#include "E01B_frequent_words.h"

set<string> FrequentWords(const string &genome, int k, int limit = -1) {
    set<string> freqPatterns;
    int *count = static_cast<int *>(calloc(genome.size(), sizeof(int)));

    for (int i = 0; i < genome.size() - k; i++) {
        string pattern = Text(genome, i, k);
        auto [cnt, ind] = PatternCount(genome, pattern);
        count[i] = cnt;
    }

    if (limit < 1) {
        int max_count = *max_element(count , count + genome.size());

        for (int i = 0; i < genome.size() - k; i++) {
            if (count[i] == max_count) {
                freqPatterns.insert(Text(genome, i, k));
            }
        }
    }
    else {
        for (int i = 0; i < genome.size() - k; i++) {
            if (count[i] == limit) {
                freqPatterns.insert(Text(genome, i, k));
            }
        }
    }

    return freqPatterns;
}

set<string> FastFrequentWords(const string &genome, int k, int limit = -1) {
    set<string> freqPatterns;
    int *freqArr = ComputeFrequencies(genome, k);

    int max_count = *max_element(freqArr , freqArr + static_cast<int>(pow(4, k)));

    if (limit < 1) {
        for (int i = 0; i < pow(4, k); i++) {
            if (freqArr[i] == max_count) {
                string pattern = NumberToPattern(i, k);
                freqPatterns.insert(pattern);
            }
        }
    }
    else {
        for (int i = 0; i < pow(4, k); i++) {
            if (freqArr[i] == limit) {
                string pattern = NumberToPattern(i, k);
                freqPatterns.insert(pattern);
            }
        }
    }

    return freqPatterns;
}

set<string> FastFrequentWordsBySorting(const string &genome, int k, int limit = -1) {
    set<string> freqPatterns;
    int *index = static_cast<int*>(calloc(genome.size() - k, sizeof(int)));
    int *count = static_cast<int*>(calloc(genome.size() - k, sizeof(int)));

    for (int i = 0; i < genome.size() - k; i++) {
        string pattern = Text(genome, i, k);
        index[i] = PatternToNumber(pattern);
        count[i] = 1;
    }

    std::sort(index, index + genome.size() - k);

    for (int i = 1; i < genome.size() - k; i++) {
        if (index[i] == index[i-1]) {
            count[i] = count[i-1] + 1;
        }
    }

    if (limit < 1) {
        int max_count = *max_element(count , count + genome.size());

        for (int i = 0; i < genome.size() - k; i++) {
            if (count[i] == max_count) {
                string pattern = NumberToPattern(index[i], k);
                freqPatterns.insert(pattern);
            }
        }
    }
    else {
        for (int i = 0; i < genome.size() - k; i++) {
            if (count[i] == limit) {
                string pattern = NumberToPattern(index[i], k);
                freqPatterns.insert(pattern);
            }
        }
    }

    return freqPatterns;
}

/*
int main() {
    string _genome;
    int _k;

    cout << "Genome: ";
    cin >> _genome;
    cout << "k: ";
    cin >> _k;

    auto freq_words = FastFrequentWordsBySorting(_genome, _k);

    for (const auto& word : freq_words) {
        cout << word << endl;
    }
}
*/
